pkg/auth/providers/saml/saml_client.go

```diff
@@ -0,0 +1,385 @@
+		}
+	}
+
+	provider := SamlProviders[name]
+
+	samlURL := configToSet.RancherAPIHost + "/v1-saml/"
+	samlURL += name
+	actURL, err := url.Parse(samlURL)
+	if err != nil {
+		return fmt.Errorf("SAML: error in parsing URL")
+	}
+
+	metadataURL := *actURL
+	metadataURL.Path = metadataURL.Path + "/saml/metadata"
+	acsURL := *actURL
+	acsURL.Path = acsURL.Path + "/saml/acs"
+
+	sp := saml.ServiceProvider{
+		Key:         privKey,
+		Certificate: cert,
+		MetadataURL: metadataURL,
+		AcsURL:      acsURL,
+		Logger:      logger.DefaultLogger,
+	}
+
+	// XML unmarshal throws an error for IdP Metadata cacheDuration field, as it's of type xml Duration. Using a separate struct for unmarshaling for now
+	idm := &IDPMetadata{}
+	if idpURL != "" {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		client := &http.Client{Transport: tr}
+		resp, err := client.Get(idpURL)
+		if err != nil {
+			return fmt.Errorf("SAML: cannot initialize saml SP, cannot get IDP Metadata  from the url %v, error %v", idpURL, err)
+		}
+		sp.IDPMetadata = &saml.EntityDescriptor{}
+		if err := xml.NewDecoder(resp.Body).Decode(idm); err != nil {
+			return fmt.Errorf("SAML: cannot initialize saml SP, cannot decode IDP Metadata xml from the config %v, error %v", configToSet, err)
+		}
+	} else if configToSet.IDPMetadataContent != "" {
+		sp.IDPMetadata = &saml.EntityDescriptor{}
+		if err := xml.NewDecoder(strings.NewReader(configToSet.IDPMetadataContent)).Decode(idm); err != nil {
+			return fmt.Errorf("SAML: cannot initialize saml SP, cannot decode IDP Metadata content from the config %v, error %v", configToSet, err)
+		}
+	} else if configToSet.IDPMetadataFilePath != "" {
+		file, err := os.Open(configToSet.IDPMetadataFilePath)
+		if err != nil {
+			return fmt.Errorf("SAML: cannot initialize saml SP, cannot read IDP Metadata file from the config %v, error %v", configToSet, err)
+		}
+		metadataReader := bufio.NewReader(file)
+		sp.IDPMetadata = &saml.EntityDescriptor{}
+		if err := xml.NewDecoder(metadataReader).Decode(idm); err != nil {
+			return fmt.Errorf("SAML: cannot initialize saml SP, cannot decode IDP Metadata xml from the config %v, error %v", configToSet, err)
+		}
+	}
+
+	sp.IDPMetadata.XMLName = idm.XMLName
+	sp.IDPMetadata.ValidUntil = idm.ValidUntil
+	sp.IDPMetadata.EntityID = idm.EntityID
+	sp.IDPMetadata.SPSSODescriptors = idm.SPSSODescriptors
+	sp.IDPMetadata.IDPSSODescriptors = idm.IDPSSODescriptors
+	provider.ServiceProvider = &sp
+
+	cookieStore := samlsp.ClientCookies{
+		ServiceProvider: &sp,
+		Name:            "token",
+		Domain:          actURL.Host,
+	}
+	provider.ClientState = &cookieStore
+
+	if name == PingName {
+		Root.Get("PingLogin").HandlerFunc(provider.HandleSamlLogin)
+		Root.Get("PingACS").HandlerFunc(provider.ServeHTTP)
+		Root.Get("PingMetadata").HandlerFunc(provider.ServeHTTP)
+	}
+	return nil
+}
+
+func PingHandlers() *mux.Router {
```

Comment:
this is called from outside of the auth packages. caller shouldnt need to know its for ping specifically.
rename as `SAMLAuthHandler` and change the return type to just http.Handler. Caller should not know implementation details of what is being returned